### ISSUE TITLE
Could support multi-line regexps with github.com/icholy/replace

### ISSUE THREAD

**Issue by tv42** [cespare/sub#11]
After a little IRC chat, @icholy wrote a neat streaming replacer. And he made it do regexps too! (As long as they're guaranteed to have matches shorter than a threshold, default 64 KiB.)

sub could use icholy/replacer to implement multi-line regexps. See flags `s` and `m`, https://github.com/google/re2/wiki/Syntax

This could be used for things like `sub -m '\n\n\n+' '\n'` to replace two or more blank lines with just one.

Background: http://doc.cat-v.org/bell_labs/structural_regexps/

(`replace` might still be buggy: https://github.com/icholy/replace/issues/2)

**Comment by tv42** [comment]
Also if I'm too much of a burden just let me know, I'll pile up these ideas and once the pile is big enough I'll end up writing code for them myself ;-)

**Comment by icholy** [comment]
The `replace.Regexp*`  functions are allowed to call the replace callback functions multiple times with the same match. (This happens when the destination buffer is too small). This will cause `sub` to print duplicates of the substitutions it made. Instead of calling the callback again, I should be saving the result on failure. So `replace` won't work here as-is.

**Comment by tv42** [comment]
Oh I didn't think about the diff output mode. With `replace`, the data at hand isn't necessarily line-aligned either!

It could do something like have a wrapper that insists on widening the "window" to the newlines preceeding and following the actually relevant data, just to output a (potentially multi-line) diff.

**Comment by tv42** [comment]
Or just use an *actual* differ and feed the original and replaced content to it? That would get you `diff -u` style context, too (at the cost of having more fluff in the output).

**Comment by icholy** [comment]
Line aligned wouldn't be enough because, if there were multiple matches on a single line, the callback would be invoked with each match separately and output duplicates. You could make an api where the callback gets all the matches in a line, but that falls apart with multi-line matches.

**Comment by tv42** [comment]
Yeah I'd personally rather just feed the result to a separate differ. I like the `diff -u` context.
